### DISCUSSION TITLE
a recommendation for good design inspiration: focus editor

### DISCUSSION THREAD

**Discussion by rahil627** [helix-editor/helix#11833] [Ideas]
recently i've been feeling helix is just... _too_ heavy. _too_ much between me and simply doing what i want. the key-bindings. the minor modes. the config. all of it. just the other day i tried this editor called [focus editor](https://github.com/focus-editor/focus) which i noted from one of j blow's streams, and i've found the experience of using it not only immensely relieving 'n refreshing, but its design choices are incredible in many tiny ways. at the minimum, i highly recommend it as a replacement for all past simpler gui text editors (notepad2, notepad++, sublime(?), etc.), but also, once you're familiar enough with a project/language in which you don't really need a lsp, even for programming. i think helix can learn quite a bit from it's incredible sane defaults, ease-of-use (learn everything in 5 minutes), and nuanced design choices.

i've only used it for an hour now, but some things i've discovered:
  - a simple welcome screen that starts you off with a handful of commands that are enough let you discover all you need to know
    - _this might come from good game design.. or doom-emacs_
  - a minimal command palette that has all the essential functions, and therefore makes it easy to discover and digest it all
    - basically a "picker" (drop-down fuzzy-search gui) instead of a "minor-mode" (pop-up menu with key-bindings) for space, some place where you can just throw all of your most commonly used functions without having to bind/remember a key for it
  - i think it has infinite undo history, even if you close a file..!
    - _note: this is written in jai, and the person who made that made Braid, so... ;)_
  - the cursor is great at jumping back to it's original position, when highlighting or viewing around
    - _note: this might be my fault, as i config'd it to always insert text at the position of the cursor (as opposed to before selection)_
  - it's simple view navigation (control+alt+arrow keys) is refreshingly good, in controls and feel, as is the feel of its scroll bar... but maybe these are terminal limitations..
  - auto-reload config upon save
  - it's initial key-bindings are minimal but very much functional
    - insert-mode could really use these, to jump between vi/kakoune-style to minimal emacs-style
  - stupid easy workspace setup (add path to config file)
    - this stupid simple idea elegantly solves the multiple workspaces problem, and prevents you from ever using `cd`
  - great little path navigation gui popup when opening file, wayyy better than windows os, lol
    - iirc: in helix, you'd have to `../../../` your way backward from the workspace path you're in...? also tab here selects the first choice, which  feels really snappy
- diff on the scroll-bar side
  - _i don't even know if helix has a scroll bar..?? the no line number life.._
- no more CRLF!
  - (and i think this app's primary target is windows os too, since jai is made for game-dev)

anyway, just a really good source of inspiration. just try it out for 5 minutes and you'll understand. but again, i'm personally guessing _some_ of its nice design choices come from the help of it being a proper gui. I don't know if its low-latency/snappiness is my just my perception (i use wezterm), but, my god... i haven't felt anything this good since notepad2 came out..!

also, rip jai language. github detects it as objective-c 🤣

**Comment by David-Else** [comment]
> auto-reload config upon save

That would be great to add to Helix and would be very easy to implement. I can't think of a reason to not have it as default, not sure why nobody thought of it before?

**Reply by rahil627** [reply]
it's yet another thing in my custom minor mode. along with `:config-open`.. 😢

**Reply by kirawi** [reply]
It was proposed, but it requires a file watcher.

**Reply by David-Else** [reply]
@kirawi Why does it need a file watcher to reload the `config.toml` when it is saved? Surely check if the file just saved was the config, if so reload it?

**Reply by kirawi** [reply]
https://github.com/helix-editor/helix/issues/8681

That doesn't work if there's more than one Helix instance. I suppose it might still be fine to just have one instance, that's for the maintainers to decide though.
